build/builder_test.go

```diff
@@ -230,3 +234,268 @@ func TestDontCountContentOfSkippedFiles(t *testing.T) {
+				},
+			},
+			expectedShardCount: 1,
+			expectedRepository: zoekt.Repository{Name: "sameName", ID: 5},
+		},
+		{
+			name:       "match on ID, not name",
+			skipReason: "We only support matching on IDS for compound shards. For normal shards, repositories wih the same name will produce shards with the same filename (which breaks our matching logic).",
+			normalShards: []normalShard{
+				{Repository: zoekt.Repository{Name: "sameName", ID: 1}},
+				{Repository: zoekt.Repository{Name: "sameName", ID: 2}},
+				{Repository: zoekt.Repository{Name: "sameName", ID: 3}},
+			},
+			compoundShards: [][]zoekt.Repository{
+				{
+					{Name: "sameName", ID: 4},
+					{Name: "sameName", ID: 5},
+					{Name: "sameName", ID: 6},
+				},
+				{
+					{Name: "sameName", ID: 7},
+					{Name: "sameName", ID: 8},
+					{Name: "sameName", ID: 9},
+				},
+			},
+			expectedShardCount: 1,
+			expectedRepository: zoekt.Repository{Name: "sameName", ID: 5},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.skipReason != "" {
+				t.Skip(tt.skipReason)
+			}
+
+			// prepare
+			indexDir := t.TempDir()
+
+			for _, s := range tt.normalShards {
+				createTestShard(t, indexDir, s.Repository, s.NumShards)
+			}
+
+			for _, repositoryGroup := range tt.compoundShards {
+				createTestCompoundShard(t, indexDir, repositoryGroup)
+			}
+
+			o := &Options{
+				IndexDir:              indexDir,
+				RepositoryDescription: tt.expectedRepository,
+			}
+			o.SetDefaults()
+
+			// run test
+			shards := o.FindAllShards()
+
+			// verify results
+			if len(shards) != tt.expectedShardCount {
+				t.Fatalf("expected %d shard(s), received %d shard(s)", tt.expectedShardCount, len(shards))
+			}
+
+			if tt.expectedShardCount > 0 {
+				for _, s := range shards {
+					// all shards should contain the metadata for the desired repository
+					repos, _, err := zoekt.ReadMetadataPathAlive(s)
+					if err != nil {
+						t.Fatalf("reading metadata from shard %q: %s", s, err)
+					}
+
+					foundRepository := false
+					for _, r := range repos {
+						if r.ID == tt.expectedRepository.ID {
+							foundRepository = true
+							break
+						}
+					}
+
+					if !foundRepository {
+						t.Errorf("shard %q doesn't contain metadata for repository %d", s, tt.expectedRepository.ID)
+					}
+				}
+			}
+		})
+	}
+}
+
+func createTestShard(t *testing.T, indexDir string, r zoekt.Repository, numShards int) {
+	t.Helper()
+
+	if err := os.MkdirAll(filepath.Dir(indexDir), 0700); err != nil {
+		t.Fatal(err)
+	}
+
+	o := Options{
+		IndexDir:              indexDir,
+		RepositoryDescription: r,
+		ShardMax:              75, // create a new shard every 75 bytes
+	}
+	o.SetDefaults()
+
+	b, err := NewBuilder(o)
+	if err != nil {
+		t.Fatalf("NewBuilder: %v", err)
+	}
+
+	if numShards == 0 {
+		// We have to make at least 1 shard.
+		numShards = 1
+	}
+
+	for i := 0; i < numShards; i++ {
+		// Create entries (file + contents) that are ~100 bytes each.
```

Comment:
This seems fairly kludgy to me, but this is the only way that I could think of to force a repository to be split across multiple shards for testing purposes.

Reply:
LGTM!